tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java

```diff
@@ -5,149 +5,175 @@
- * The code was written by Allan Heydon
+ * so that most operations do not require a disk access.
+ *
+ * <p>This class is not thread safe.
+ *
+ * @author Allan Heydon
  */
-
-import java.io.File;
-import java.io.IOException;
-import java.math.BigInteger;
-
 public final class BufferedRandomAccessFile extends java.io.RandomAccessFile {
-	// Increase 8k buffer to match modern day hardware? No!!! The implementation
-	// discards the whole buffer each time it seeks to a position outside the
-	// current buffer. It then creates a new buffer, which takes proportionally
-	// longer with larger buffer sizes.
-	static final int LogBuffSz = 13; // 8K buffer
+    // Increase 8k buffer to match modern day hardware? No!!! The implementation
+    // discards the whole buffer each time it seeks to a position outside the
+    // current buffer. It then creates a new buffer, which takes proportionally
+    // longer with larger buffer sizes.
+    static final int LogBuffSz = 13; // 8K buffer
     public static final int BuffSz = (1 << LogBuffSz);
     static final int BuffMask = ~(BuffSz - 1);
 
     /* This implementation is based on the buffer implementation in
-       Modula-3's "Rd", "Wr", "RdClass", and "WrClass" interfaces. */
+       Modula-3's "Rd", "Wr", "RdClass", and "WrClass" interfaces,
+       with substantial modifications. */
+
+    /**
+     * True if unflushed bytes exist.
+     * @see #flushBuffer()
+     */
     private boolean dirty;  // true iff unflushed bytes exist
+
     // SZ Feb 24, 2009: never read
     // private boolean closed; // true iff the file is closed
-    private long curr;      // current position in file
-    private long lo, hi;    // bounds on characters in "buff"
-    private byte[] buff;    // local buffer
-    private long maxHi;     // this.lo + this.buff.length
-    private boolean hitEOF; // buffer contains last file block?
-    private long diskPos;   // disk position
-	private long mark;
-    
-    private static final Object mu = new Object(); // protects the following fields
+
+    /** Current length of the file. */
+    private long length;
+
+    /** Current position in the file. */
+    private long curr;
+
+    /** Offset of <code>buff[0]</code> in the file. */
+    private long lo;
+
+    /** Cache of the bytes at indexes <code>[lo, min(lo + BuffSz, length))</code>. */
+    private final byte[] buff;
+
+    /**
+     * Cached value of <code>super.{@link #getFilePointer()}</code>.  Used to avoid
+     * some calls to <code>super.{@link #seek(long)}</code>.
+     */
+    private long diskPos;
+
+    /**
+     * Value of the mark.
+     * @see #mark()
+     * @see #reset()
+     */
+    private long mark;
+
+    /** Protects {@link #availBuffs} and {@link #numAvailBuffs}. */
+    private static final Object mu = new Object();
     private static byte[][] availBuffs = new byte[100][];
     private static int numAvailBuffs = 0;
 
     /* To describe the above fields, we introduce the following
```

Comment:
Before the spec existed, this is clearly the right place for this comment to exist.  However, given that there is now a formal spec for this module, it seems like it might be appealing to put the bulk of this description into the spec, and then refer the reader to the spec here in this file.

Reply:
I am actually going to leave it here, since some of the other comments in this file refer to specific invariants by name, and I think it's helpful to have all that in one file.

But, I will definitely add a note here pointing to the TLA+ formalization.